## FLIGHT PREDICTION DATA SET EDA AND FEATURE ENGINEERING 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
# HERE WE HAVE TESTING AND TRAINING DATA GIVEN TO US 
# SO IMPORTING THE  DATA SETS

In [52]:
train_data = pd.read_excel('Data_Train.xlsx')
train_data.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [53]:
test_data = pd.read_excel('Test_set.xlsx')
test_data.head()

Airline Date_of_Journey    Source Destination            Route  \
0        Jet Airways       6/06/2019     Delhi      Cochin  DEL → BOM → COK   
1             IndiGo      12/05/2019   Kolkata    Banglore  CCU → MAA → BLR   
2        Jet Airways      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
3  Multiple carriers      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
4           Air Asia      24/06/2019  Banglore       Delhi        BLR → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops              Additional_Info  
0    17:30  04:25 07 Jun  10h 55m      1 stop                      No info  
1    06:20         10:20       4h      1 stop                      No info  
2    19:15  19:00 22 May  23h 45m      1 stop  In-flight meal not included  
3    08:00         21:00      13h      1 stop                      No info  
4    23:55  02:45 25 Jun   2h 50m    non-stop                      No info

In [54]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [55]:
train_data.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

## CHECKING FOR NULL VALUES 
Here we have only one null vlaue in both route and total stops 
as the data set is having total of 10683 entries , 
we can drop one entry from the data set to eliminate the null values

In [56]:
train_data.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

In [57]:
train_data.shape

(10683, 11)

In [58]:
train_data[train_data['Route'].isnull() == True ]       # NULL Value present in the below given entry

Airline Date_of_Journey Source Destination Route Dep_Time  \
9039  Air India       6/05/2019  Delhi      Cochin   NaN    09:45   

      Arrival_Time Duration Total_Stops Additional_Info  Price  
9039  09:25 07 May  23h 40m         NaN         No info   7480

In [59]:
train_data.dropna(axis=0, inplace=True)

In [60]:
train_data.shape

(10682, 11)

In [61]:
train_data.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              0
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        0
Additional_Info    0
Price              0
dtype: int64

## COMBINING THE TRAINING AND TESTING DATA TO PERFORM EDA

In [62]:
data = train_data._append(test_data)

In [63]:
data.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info    Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897.0  
1    05:50         13:15   7h 25m     2 stops         No info   7662.0  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882.0  
3    18:05         23:30   5h 25m      1 stop         No info   6218.0  
4    16:50         21:35   4h 45m      1 stop         No info  13302.0

In [64]:
data.tail()

Airline Date_of_Journey   Source Destination            Route  \
2666          Air India       6/06/2019  Kolkata    Banglore  CCU → DEL → BLR   
2667             IndiGo      27/03/2019  Kolkata    Banglore        CCU → BLR   
2668        Jet Airways       6/03/2019    Delhi      Cochin  DEL → BOM → COK   
2669          Air India       6/03/2019    Delhi      Cochin  DEL → BOM → COK   
2670  Multiple carriers      15/06/2019    Delhi      Cochin  DEL → BOM → COK   

     Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
2666    20:30  20:25 07 Jun  23h 55m      1 stop         No info    NaN  
2667    14:20         16:55   2h 35m    non-stop         No info    NaN  
2668    21:50  04:25 07 Mar   6h 35m      1 stop         No info    NaN  
2669    04:00         19:15  15h 15m      1 stop         No info    NaN  
2670    04:55         19:15  14h 20m      1 stop         No info    NaN

## EXPLORING DIFFERENT FEATURES OF THE DATA SET

In [65]:
data.Source.value_counts()

Source
Delhi       5681
Kolkata     3581
Banglore    2752
Mumbai       883
Chennai      456
Name: count, dtype: int64

In [66]:
data.Destination.value_counts()

Destination
Cochin       5681
Banglore     3581
Delhi        1582
New Delhi    1170
Hyderabad     883
Kolkata       456
Name: count, dtype: int64

In [67]:
data.Total_Stops.value_counts()

Total_Stops
1 stop      7056
non-stop    4340
2 stops     1899
3 stops       56
4 stops        2
Name: count, dtype: int64

In [68]:
data.Airline.value_counts()

Airline
Jet Airways                          4746
IndiGo                               2564
Air India                            2191
Multiple carriers                    1543
SpiceJet                             1026
Vistara                               608
Air Asia                              405
GoAir                                 240
Multiple carriers Premium economy      16
Jet Airways Business                    8
Vistara Premium economy                 5
Trujet                                  1
Name: count, dtype: int64

In [69]:
data.Additional_Info.value_counts()

Additional_Info
No info                         10492
In-flight meal not included      2426
No check-in baggage included      396
1 Long layover                     20
Change airports                     8
Business class                      5
No Info                             3
1 Short layover                     1
Red-eye flight                      1
2 Long layover                      1
Name: count, dtype: int64

## SPLITTING DATE OF JOURNEY INTO DAY MONTH AND YEAR IN INTEGER FORMAT

In [70]:
day = []
month = []
year = []

for date in data['Date_of_Journey']:
    date = str(date).split('/')
    day.append(int(date[0]))
    month.append(int(date[1]))
    year.append(int(date[2]))

data.insert(loc=1, column='JR_Day', value=day)
data.insert(loc=2, column='JR_Month', value=month)
data.insert(loc=3, column='JR_Year', value=year)

In [71]:
data.head()

Airline  JR_Day  JR_Month  JR_Year Date_of_Journey    Source  \
0       IndiGo      24         3     2019      24/03/2019  Banglore   
1    Air India       1         5     2019       1/05/2019   Kolkata   
2  Jet Airways       9         6     2019       9/06/2019     Delhi   
3       IndiGo      12         5     2019      12/05/2019   Kolkata   
4       IndiGo       1         3     2019      01/03/2019  Banglore   

  Destination                  Route Dep_Time  Arrival_Time Duration  \
0   New Delhi              BLR → DEL    22:20  01:10 22 Mar   2h 50m   
1    Banglore  CCU → IXR → BBI → BLR    05:50         13:15   7h 25m   
2      Cochin  DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h   
3    Banglore        CCU → NAG → BLR    18:05         23:30   5h 25m   
4   New Delhi        BLR → NAG → DEL    16:50         21:35   4h 45m   

  Total_Stops Additional_Info    Price  
0    non-stop         No info   3897.0  
1     2 stops         No info   7662.0  
2     2 stops         No info  13882.0  
3      1 stop         No info   6218.0  
4      1 stop         No info  13302.0

In [72]:
data.drop(columns=['Date_of_Journey'], axis=1,inplace=True)

In [73]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  \
0       IndiGo      24         3     2019  Banglore   New Delhi   
1    Air India       1         5     2019   Kolkata    Banglore   
2  Jet Airways       9         6     2019     Delhi      Cochin   
3       IndiGo      12         5     2019   Kolkata    Banglore   
4       IndiGo       1         3     2019  Banglore   New Delhi   

                   Route Dep_Time  Arrival_Time Duration Total_Stops  \
0              BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1  CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2  DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3        CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4        BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   

  Additional_Info    Price  
0         No info   3897.0  
1         No info   7662.0  
2         No info  13882.0  
3         No info   6218.0  
4         No info  13302.0

## SPLITTING THE ARRIVAL TIME AND REMOVING THE UNECESSARY DATE 

In [74]:
at = []
for time in data['Arrival_Time']:
    time = str(time).split(' ')[0]
    at.append(time)



In [75]:
data.drop(columns=['Arrival_Time'],axis=1, inplace=True)

In [76]:
data.insert(loc=8,column='Arrival_Time', value=at)

In [77]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  \
0       IndiGo      24         3     2019  Banglore   New Delhi   
1    Air India       1         5     2019   Kolkata    Banglore   
2  Jet Airways       9         6     2019     Delhi      Cochin   
3       IndiGo      12         5     2019   Kolkata    Banglore   
4       IndiGo       1         3     2019  Banglore   New Delhi   

                   Route Dep_Time Arrival_Time Duration Total_Stops  \
0              BLR → DEL    22:20        01:10   2h 50m    non-stop   
1  CCU → IXR → BBI → BLR    05:50        13:15   7h 25m     2 stops   
2  DEL → LKO → BOM → COK    09:25        04:25      19h     2 stops   
3        CCU → NAG → BLR    18:05        23:30   5h 25m      1 stop   
4        BLR → NAG → DEL    16:50        21:35   4h 45m      1 stop   

  Additional_Info    Price  
0         No info   3897.0  
1         No info   7662.0  
2         No info  13882.0  
3         No info   6218.0  
4         No info  13302.0

## NOW SPLITTING ARRIVAL AND DEPARTURE TIME INTO HOURS AND MINUTES

In [78]:
arr_hrs = data['Arrival_Time'].str.split(':').str[0].astype(int)
arr_mins = data['Arrival_Time'].str.split(':').str[1].astype(int)
dep_hrs = data['Dep_Time'].str.split(':').str[0].astype(int)
dep_mins = data['Dep_Time'].str.split(':').str[1].astype(int)

In [79]:
data.drop(columns=['Arrival_Time','Dep_Time'], axis=1, inplace=True)

In [80]:
data.insert(loc=7,column='Dep_Hours', value=dep_hrs)
data.insert(loc=8,column='Dep_Mins', value=dep_mins)
data.insert(loc=9,column='Arrival_Hours', value=arr_hrs)
data.insert(loc=10,column='Arrival_Min', value=arr_mins)

In [81]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  \
0       IndiGo      24         3     2019  Banglore   New Delhi   
1    Air India       1         5     2019   Kolkata    Banglore   
2  Jet Airways       9         6     2019     Delhi      Cochin   
3       IndiGo      12         5     2019   Kolkata    Banglore   
4       IndiGo       1         3     2019  Banglore   New Delhi   

                   Route  Dep_Hours  Dep_Mins  Arrival_Hours  Arrival_Min  \
0              BLR → DEL         22        20              1           10   
1  CCU → IXR → BBI → BLR          5        50             13           15   
2  DEL → LKO → BOM → COK          9        25              4           25   
3        CCU → NAG → BLR         18         5             23           30   
4        BLR → NAG → DEL         16        50             21           35   

  Duration Total_Stops Additional_Info    Price  
0   2h 50m    non-stop         No info   3897.0  
1   7h 25m     2 stops         No info   7662.0  
2      19h     2 stops         No info  13882.0  
3   5h 25m      1 stop         No info   6218.0  
4   4h 45m      1 stop         No info  13302.0

In [82]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13353 entries, 0 to 2670
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13353 non-null  object 
 1   JR_Day           13353 non-null  int64  
 2   JR_Month         13353 non-null  int64  
 3   JR_Year          13353 non-null  int64  
 4   Source           13353 non-null  object 
 5   Destination      13353 non-null  object 
 6   Route            13353 non-null  object 
 7   Dep_Hours        13353 non-null  int64  
 8   Dep_Mins         13353 non-null  int64  
 9   Arrival_Hours    13353 non-null  int64  
 10  Arrival_Min      13353 non-null  int64  
 11  Duration         13353 non-null  object 
 12  Total_Stops      13353 non-null  object 
 13  Additional_Info  13353 non-null  object 
 14  Price            10682 non-null  float64
dtypes: float64(1), int64(7), object(7)
memory usage: 1.6+ MB


## CONVERTING THE DATA TYPE OF TOTAL STOPS FROM OBJECT TO INTEGER

In [83]:
data['Total_Stops'] = data['Total_Stops'].map(
    {
        '1 stop'  : 1,
        'non-stop'  : 0,
        '2 stops'  : 2,
        '3 stops'  : 3,
        '4 stops'  : 4
    }
)

In [84]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  \
0       IndiGo      24         3     2019  Banglore   New Delhi   
1    Air India       1         5     2019   Kolkata    Banglore   
2  Jet Airways       9         6     2019     Delhi      Cochin   
3       IndiGo      12         5     2019   Kolkata    Banglore   
4       IndiGo       1         3     2019  Banglore   New Delhi   

                   Route  Dep_Hours  Dep_Mins  Arrival_Hours  Arrival_Min  \
0              BLR → DEL         22        20              1           10   
1  CCU → IXR → BBI → BLR          5        50             13           15   
2  DEL → LKO → BOM → COK          9        25              4           25   
3        CCU → NAG → BLR         18         5             23           30   
4        BLR → NAG → DEL         16        50             21           35   

  Duration  Total_Stops Additional_Info    Price  
0   2h 50m            0         No info   3897.0  
1   7h 25m            2         No info   7662.0  
2      19h            2         No info  13882.0  
3   5h 25m            1         No info   6218.0  
4   4h 45m            1         No info  13302.0

In [85]:
data.drop(columns=['Route'],axis=1,inplace=True)

In [86]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0       IndiGo      24         3     2019  Banglore   New Delhi         22   
1    Air India       1         5     2019   Kolkata    Banglore          5   
2  Jet Airways       9         6     2019     Delhi      Cochin          9   
3       IndiGo      12         5     2019   Kolkata    Banglore         18   
4       IndiGo       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min Duration  Total_Stops Additional_Info  \
0        20              1           10   2h 50m            0         No info   
1        50             13           15   7h 25m            2         No info   
2        25              4           25      19h            2         No info   
3         5             23           30   5h 25m            1         No info   
4        50             21           35   4h 45m            1         No info   

     Price  
0   3897.0  
1   7662.0  
2  13882.0  
3   6218.0  
4  13302.0

In [93]:
data.shape

(13353, 14)

## CONVERTING THE DURATION INTO MINS

In [ ]:
dur = data['Duration']
min_dur = []
for d in dur:
    d = str(d.replace('h','').replace('m', ''))
    d = d.split(' ')

    if len(d) == 1:
        d = int(d[0])
        if d < 60 :
            pass
        else:
            d = d*60
    else:
        d = int(d[0])*60 + int(d[1])
    min_dur.append(d)


13353


In [104]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0       IndiGo      24         3     2019  Banglore   New Delhi         22   
1    Air India       1         5     2019   Kolkata    Banglore          5   
2  Jet Airways       9         6     2019     Delhi      Cochin          9   
3       IndiGo      12         5     2019   Kolkata    Banglore         18   
4       IndiGo       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min Duration  Total_Stops Additional_Info  \
0        20              1           10   2h 50m            0         No info   
1        50             13           15   7h 25m            2         No info   
2        25              4           25      19h            2         No info   
3         5             23           30   5h 25m            1         No info   
4        50             21           35   4h 45m            1         No info   

     Price  
0   3897.0  
1   7662.0  
2  13882.0  
3   6218.0  
4  13302.0

In [106]:
data.insert(loc=11, column='Duration in Minutes', value=min_dur)


In [107]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0       IndiGo      24         3     2019  Banglore   New Delhi         22   
1    Air India       1         5     2019   Kolkata    Banglore          5   
2  Jet Airways       9         6     2019     Delhi      Cochin          9   
3       IndiGo      12         5     2019   Kolkata    Banglore         18   
4       IndiGo       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min Duration  Duration in Minutes  \
0        20              1           10   2h 50m                  170   
1        50             13           15   7h 25m                  445   
2        25              4           25      19h                   19   
3         5             23           30   5h 25m                  325   
4        50             21           35   4h 45m                  285   

   Total_Stops Additional_Info    Price  
0            0         No info   3897.0  
1            2         No info   7662.0  
2            2         No info  13882.0  
3            1         No info   6218.0  
4            1         No info  13302.0

In [108]:
data.drop(columns=['Duration'],axis=1,inplace=True)

In [109]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0       IndiGo      24         3     2019  Banglore   New Delhi         22   
1    Air India       1         5     2019   Kolkata    Banglore          5   
2  Jet Airways       9         6     2019     Delhi      Cochin          9   
3       IndiGo      12         5     2019   Kolkata    Banglore         18   
4       IndiGo       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min  Duration in Minutes  Total_Stops  \
0        20              1           10                  170            0   
1        50             13           15                  445            2   
2        25              4           25                   19            2   
3         5             23           30                  325            1   
4        50             21           35                  285            1   

  Additional_Info    Price  
0         No info   3897.0  
1         No info   7662.0  
2         No info  13882.0  
3         No info   6218.0  
4         No info  13302.0

## LABEL ENCODING THE REMAINING CATEGORICAL DATESETS : SOURCE,DESTINATION , AIRLINES, ADDITIONAL INFO

In [111]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [112]:
data['Airline'] = le.fit_transform(data['Airline'])

In [113]:
data.head()

Airline  JR_Day  JR_Month  JR_Year    Source Destination  Dep_Hours  \
0        3      24         3     2019  Banglore   New Delhi         22   
1        1       1         5     2019   Kolkata    Banglore          5   
2        4       9         6     2019     Delhi      Cochin          9   
3        3      12         5     2019   Kolkata    Banglore         18   
4        3       1         3     2019  Banglore   New Delhi         16   

   Dep_Mins  Arrival_Hours  Arrival_Min  Duration in Minutes  Total_Stops  \
0        20              1           10                  170            0   
1        50             13           15                  445            2   
2        25              4           25                   19            2   
3         5             23           30                  325            1   
4        50             21           35                  285            1   

  Additional_Info    Price  
0         No info   3897.0  
1         No info   7662.0  
2         No info  13882.0  
3         No info   6218.0  
4         No info  13302.0

In [115]:
data.Airline.unique()

array([ 3,  1,  4,  8,  6,  2, 10,  0, 11,  5,  7,  9])

In [116]:
data['Source'] = le.fit_transform(data['Source'])
data['Destination'] = le.fit_transform(data['Destination'])
data['Additional_Info'] = le.fit_transform(data['Additional_Info'])

In [117]:
data.head()

Airline  JR_Day  JR_Month  JR_Year  Source  Destination  Dep_Hours  \
0        3      24         3     2019       0            5         22   
1        1       1         5     2019       3            0          5   
2        4       9         6     2019       2            1          9   
3        3      12         5     2019       3            0         18   
4        3       1         3     2019       0            5         16   

   Dep_Mins  Arrival_Hours  Arrival_Min  Duration in Minutes  Total_Stops  \
0        20              1           10                  170            0   
1        50             13           15                  445            2   
2        25              4           25                   19            2   
3         5             23           30                  325            1   
4        50             21           35                  285            1   

   Additional_Info    Price  
0                8   3897.0  
1                8   7662.0  
2                8  13882.0  
3                8   6218.0  
4                8  13302.0

In [118]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13353 entries, 0 to 2670
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Airline              13353 non-null  int64  
 1   JR_Day               13353 non-null  int64  
 2   JR_Month             13353 non-null  int64  
 3   JR_Year              13353 non-null  int64  
 4   Source               13353 non-null  int64  
 5   Destination          13353 non-null  int64  
 6   Dep_Hours            13353 non-null  int64  
 7   Dep_Mins             13353 non-null  int64  
 8   Arrival_Hours        13353 non-null  int64  
 9   Arrival_Min          13353 non-null  int64  
 10  Duration in Minutes  13353 non-null  int64  
 11  Total_Stops          13353 non-null  int64  
 12  Additional_Info      13353 non-null  int64  
 13  Price                10682 non-null  float64
dtypes: float64(1), int64(13)
memory usage: 1.5 MB
